File: ts/indicators/john-ehlers/hilbert-transformer-instantaneous-trend-line/hilbert-transformer-instantaneous-trend-line.ts
import { buildMetadata } from '../../core/build-metadata';
import { Bar } from '../../../entities/bar';
import { BarComponent, barComponentValue } from '../../../entities/bar-component';
import { Quote } from '../../../entities/quote';
import { DefaultQuoteComponent, quoteComponentValue } from '../../../entities/quote-component';
import { Scalar } from '../../../entities/scalar';
import { Trade } from '../../../entities/trade';
import { DefaultTradeComponent, tradeComponentValue } from '../../../entities/trade-component';
import { componentTripleMnemonic } from '../../core/component-triple-mnemonic';
import { Indicator } from '../../core/indicator';
import { IndicatorMetadata } from '../../core/indicator-metadata';
import { IndicatorOutput } from '../../core/indicator-output';
import { IndicatorIdentifier } from '../../core/indicator-identifier';
import { HilbertTransformerCycleEstimator } from '../hilbert-transformer/cycle-estimator';
import { HilbertTransformerCycleEstimatorType } from '../hilbert-transformer/cycle-estimator-type';
import { createEstimator, estimatorMoniker } from '../hilbert-transformer/common';
import { HilbertTransformerInstantaneousTrendLineParams } from './params';

const DEFAULT_ALPHA_EMA_PERIOD_ADDITIONAL = 0.33;
const DEFAULT_SMOOTHING_LENGTH = 4;
const DEFAULT_ALPHA_EMA_QI = 0.2;
const DEFAULT_ALPHA_EMA_PERIOD = 0.2;
// MBST's DominantCyclePeriod default warm-up is MaxPeriod * 2 = 100.
const DEFAULT_WARM_UP_PERIOD = 100;
const DEFAULT_TREND_LINE_SMOOTHING_LENGTH = 4;
const DEFAULT_CYCLE_PART_MULTIPLIER = 1.0;
const MAX_CYCLE_PART_MULTIPLIER = 10.0;

/** __Hilbert Transformer Instantaneous Trend Line__ (Ehlers) computes a smoothed trend line
 * derived from averaging over a window whose length tracks the smoothed dominant cycle period
 * produced by a Hilbert transformer cycle estimator.
 *
 * It exposes two outputs:
 *
 *	- Value: the instantaneous trend line value, a length-`trendLineSmoothingLength` WMA of
 *	  simple averages over the last `round(smoothedPeriod·cyclePartMultiplier)` raw input samples.
 *	- DominantCyclePeriod: the additionally EMA-smoothed dominant cycle period.
 *
 * Reference:
 *
 *	John Ehlers, Rocket Science for Traders, Wiley, 2001, 0471405671, pp 107-112.
 */
export class HilbertTransformerInstantaneousTrendLine implements Indicator {
  private readonly htce: HilbertTransformerCycleEstimator;
  private readonly alphaEmaPeriodAdditional: number;
  private readonly oneMinAlphaEmaPeriodAdditional: number;
  private readonly cyclePartMultiplier: number;
  private readonly trendLineSmoothingLength: number;
  private readonly coeff0: number;
  private readonly coeff1: number;
  private readonly coeff2: number;
  private readonly coeff3: number;
  private readonly input: number[];
  private readonly inputLength: number;
  private readonly inputLengthMin1: number;
  private smoothedPeriod = 0;
  private value = Number.NaN;
  private average1 = 0;
  private average2 = 0;
  private average3 = 0;
  private primed = false;

  private readonly mnemonicValue: string;
  private readonly descriptionValue: string;
  private readonly mnemonicDCP: string;
  private readonly descriptionDCP: string;

  private readonly barComponentFunc: (bar: Bar) => number;
  private readonly quoteComponentFunc: (quote: Quote) => number;
  private readonly tradeComponentFunc: (trade: Trade) => number;

  /** Creates an instance using default parameters (α=0.33, trendLineSmoothingLength=4,
   * cyclePartMultiplier=1.0, HomodyneDiscriminator cycle estimator with smoothingLength=4,
   * αq=0.2, αp=0.2, warmUpPeriod=100, BarComponent.Median). */
  public static default(): HilbertTransformerInstantaneousTrendLine {
    return new HilbertTransformerInstantaneousTrendLine({
      alphaEmaPeriodAdditional: DEFAULT_ALPHA_EMA_PERIOD_ADDITIONAL,
      trendLineSmoothingLength: DEFAULT_TREND_LINE_SMOOTHING_LENGTH,
      cyclePartMultiplier: DEFAULT_CYCLE_PART_MULTIPLIER,
      estimatorType: HilbertTransformerCycleEstimatorType.HomodyneDiscriminator,
      estimatorParams: {
        smoothingLength: DEFAULT_SMOOTHING_LENGTH,
        alphaEmaQuadratureInPhase: DEFAULT_ALPHA_EMA_QI,
        alphaEmaPeriod: DEFAULT_ALPHA_EMA_PERIOD,
        warmUpPeriod: DEFAULT_WARM_UP_PERIOD,
      },
    });
  }

  /** Creates an instance based on the given parameters. */
  public static fromParams(
    params: HilbertTransformerInstantaneousTrendLineParams,
  ): HilbertTransformerInstantaneousTrendLine {
    return new HilbertTransformerInstantaneousTrendLine(params);
  }

  private constructor(params: HilbertTransformerInstantaneousTrendLineParams) {
    const alpha = params.alphaEmaPeriodAdditional;
    if (alpha <= 0 || alpha > 1) {
      throw new Error(
        'invalid hilbert transformer instantaneous trend line parameters: '
          + 'α for additional smoothing should be in range (0, 1]',
      );
    }

    const tlsl = params.trendLineSmoothingLength ?? DEFAULT_TREND_LINE_SMOOTHING_LENGTH;
    if (tlsl < 2 || tlsl > 4 || !Number.isInteger(tlsl)) {
      throw new Error(
        'invalid hilbert transformer instantaneous trend line parameters: '
          + 'trend line smoothing length should be 2, 3, or 4',
      );
    }

    const cpm = params.cyclePartMultiplier ?? DEFAULT_CYCLE_PART_MULTIPLIER;
    if (cpm <= 0 || cpm > MAX_CYCLE_PART_MULTIPLIER) {
      throw new Error(
        'invalid hilbert transformer instantaneous trend line parameters: '
          + 'cycle part multiplier should be in range (0, 10]',
      );
    }

    this.alphaEmaPeriodAdditional = alpha;
    this.oneMinAlphaEmaPeriodAdditional = 1 - alpha;
    this.cyclePartMultiplier = cpm;
    this.trendLineSmoothingLength = tlsl;

    // Default to BarComponent.Median (MBST default; always shown in mnemonic).
    const bc = params.barComponent ?? BarComponent.Median;
    const qc = params.quoteComponent ?? DefaultQuoteComponent;
    const tc = params.tradeComponent ?? DefaultTradeComponent;

    this.barComponentFunc = barComponentValue(bc);
    this.quoteComponentFunc = quoteComponentValue(qc);
    this.tradeComponentFunc = tradeComponentValue(tc);

    this.htce = createEstimator(params.estimatorType, params.estimatorParams);

    const effectiveType = params.estimatorType ?? HilbertTransformerCycleEstimatorType.HomodyneDiscriminator;
    let em = '';
    const isDefaultHd = effectiveType === HilbertTransformerCycleEstimatorType.HomodyneDiscriminator
      && this.htce.smoothingLength === DEFAULT_SMOOTHING_LENGTH
      && this.htce.alphaEmaQuadratureInPhase === DEFAULT_ALPHA_EMA_QI
      && this.htce.alphaEmaPeriod === DEFAULT_ALPHA_EMA_PERIOD;
    if (!isDefaultHd) {
      const moniker = estimatorMoniker(effectiveType, this.htce);
      if (moniker.length > 0) {
        em = ', ' + moniker;
      }
    }

    const cm = componentTripleMnemonic(bc, qc, tc);
    const a = alpha.toFixed(3);
    const c = cpm.toFixed(3);

    this.mnemonicValue = `htitl(${a}, ${tlsl}, ${c}${em}${cm})`;
    this.mnemonicDCP = `dcp(${a}${em}${cm})`;

    this.descriptionValue = 'Hilbert transformer instantaneous trend line ' + this.mnemonicValue;
    this.descriptionDCP = 'Dominant cycle period ' + this.mnemonicDCP;

    // WMA coefficients.
    let c0 = 0, c1 = 0, c2 = 0, c3 = 0;
    if (tlsl === 2) {
      c0 = 2 / 3;
      c1 = 1 / 3;
    } else if (tlsl === 3) {
      c0 = 3 / 6;
      c1 = 2 / 6;
      c2 = 1 / 6;
    } else { // tlsl === 4
      c0 = 4 / 10;
      c1 = 3 / 10;
      c2 = 2 / 10;
      c3 = 1 / 10;
    }
    this.coeff0 = c0;
    this.coeff1 = c1;
    this.coeff2 = c2;
    this.coeff3 = c3;

    const maxPeriod = this.htce.maxPeriod;
    this.input = new Array<number>(maxPeriod).fill(0);
    this.inputLength = maxPeriod;
    this.inputLengthMin1 = maxPeriod - 1;
  }

  /** Indicates whether the indicator is primed. */
  public isPrimed(): boolean { return this.primed; }

  /** Describes the output data of the indicator. */
  public metadata(): IndicatorMetadata {
    return buildMetadata(
      IndicatorIdentifier.HilbertTransformerInstantaneousTrendLine,
      this.mnemonicValue,
      this.descriptionValue,
      [
        { mnemonic: this.mnemonicValue, description: this.descriptionValue },
        { mnemonic: this.mnemonicDCP, description: this.descriptionDCP },
      ],
    );
  }

  /** Updates the indicator given the next sample value. Returns the pair
   * (value, period). Returns (NaN, NaN) if not yet primed. */
  public update(sample: number): [number, number] {
    if (Number.isNaN(sample)) {
      return [sample, sample];
    }

    this.htce.update(sample);
    this.pushInput(sample);

    if (this.primed) {
      this.smoothedPeriod = this.alphaEmaPeriodAdditional * this.htce.period
        + this.oneMinAlphaEmaPeriodAdditional * this.smoothedPeriod;
      const average = this.calculateAverage();
      this.value = this.coeff0 * average
        + this.coeff1 * this.average1
        + this.coeff2 * this.average2
        + this.coeff3 * this.average3;
      this.average3 = this.average2;
      this.average2 = this.average1;
      this.average1 = average;
      return [this.value, this.smoothedPeriod];
    }

    if (this.htce.primed) {
      this.primed = true;
      this.smoothedPeriod = this.htce.period;
      const average = this.calculateAverage();
      this.value = average;
      this.average1 = average;
      this.average2 = average;
      this.average3 = average;
      return [this.value, this.smoothedPeriod];
    }

    return [Number.NaN, Number.NaN];
  }

  /** Updates an indicator given the next scalar sample. */
  public updateScalar(sample: Scalar): IndicatorOutput {
    return this.updateEntity(sample.time, sample.value);
  }

  /** Updates an indicator given the next bar sample. */
  public updateBar(sample: Bar): IndicatorOutput {
    return this.updateEntity(sample.time, this.barComponentFunc(sample));
  }

  /** Updates an indicator given the next quote sample. */
  public updateQuote(sample: Quote): IndicatorOutput {
    return this.updateEntity(sample.time, this.quoteComponentFunc(sample));
  }

  /** Updates an indicator given the next trade sample. */
  public updateTrade(sample: Trade): IndicatorOutput {
    return this.updateEntity(sample.time, this.tradeComponentFunc(sample));
  }

  private updateEntity(time: Date, sample: number): IndicatorOutput {
    const [value, period] = this.update(sample);

    const sv = new Scalar();
    sv.time = time;
    sv.value = value;

    const sp = new Scalar();
    sp.time = time;
    sp.value = period;

    return [sv, sp];
  }

  private pushInput(value: number): void {
    for (let i = this.inputLengthMin1; i > 0; i--) {
      this.input[i] = this.input[i - 1];
    }
    this.input[0] = value;
  }

  private calculateAverage(): number {
    // Compute simple average over a window tracking the smoothed dominant cycle period.
    let length = Math.floor(this.smoothedPeriod * this.cyclePartMultiplier + 0.5);
    if (length > this.inputLength) {
      length = this.inputLength;
    } else if (length < 1) {
      length = 1;
    }

    let sum = 0;
    for (let i = 0; i < length; i++) {
      sum += this.input[i];
    }
    return sum / length;
  }
}

File: go/indicators/johnehlers/hilberttransformerinstantaneoustrendline/hilberttransformerinstantaneoustrendline.go
package hilberttransformerinstantaneoustrendline

//nolint: gofumpt
import (
	"fmt"
	"math"
	"sync"
	"time"

	"zpano/entities"
	"zpano/indicators/core"
	"zpano/indicators/johnehlers/hilberttransformer"
)

// HilbertTransformerInstantaneousTrendLine is Ehlers' Instantaneous Trend Line indicator
// built on top of a Hilbert transformer cycle estimator.
//
// It exposes two outputs:
//
//   - Value: the instantaneous trend line value, computed as a WMA of simple averages
//     over windows whose length tracks the smoothed dominant cycle period.
//   - DominantCyclePeriod: the additionally EMA-smoothed dominant cycle period.
//
// Reference:
// John Ehlers, Rocket Science for Traders, Wiley, 2001, 0471405671, pp 107-112.
type HilbertTransformerInstantaneousTrendLine struct {
	mu                             sync.RWMutex
	mnemonic                       string
	description                    string
	mnemonicDCP                    string
	descriptionDCP                 string
	htce                           hilberttransformer.CycleEstimator
	alphaEmaPeriodAdditional       float64
	oneMinAlphaEmaPeriodAdditional float64
	cyclePartMultiplier            float64
	trendLineSmoothingLength       int
	coeff0                         float64
	coeff1                         float64
	coeff2                         float64
	coeff3                         float64
	smoothedPeriod                 float64
	value                          float64
	average1                       float64
	average2                       float64
	average3                       float64
	input                          []float64
	inputLength                    int
	inputLengthMin1                int
	primed                         bool
	barFunc                        entities.BarFunc
	quoteFunc                      entities.QuoteFunc
	tradeFunc                      entities.TradeFunc
}

// NewHilbertTransformerInstantaneousTrendLineDefault returns an instance of the indicator
// created using default values of the parameters.
func NewHilbertTransformerInstantaneousTrendLineDefault() (*HilbertTransformerInstantaneousTrendLine, error) {
	const (
		smoothingLength           = 4
		alphaEmaQuadratureInPhase = 0.2
		alphaEmaPeriod            = 0.2
		alphaEmaPeriodAdditional  = 0.33
		warmUpPeriod              = 100
		trendLineSmoothingLength  = 4
		cyclePartMultiplier       = 1.0
	)

	return newHilbertTransformerInstantaneousTrendLine(
		hilberttransformer.HomodyneDiscriminator,
		&hilberttransformer.CycleEstimatorParams{
			SmoothingLength:           smoothingLength,
			AlphaEmaQuadratureInPhase: alphaEmaQuadratureInPhase,
			AlphaEmaPeriod:            alphaEmaPeriod,
			WarmUpPeriod:              warmUpPeriod,
		},
		alphaEmaPeriodAdditional,
		trendLineSmoothingLength,
		cyclePartMultiplier,
		0, 0, 0)
}

// NewHilbertTransformerInstantaneousTrendLineParams returns an instance of the indicator
// created using supplied parameters.
func NewHilbertTransformerInstantaneousTrendLineParams(
	p *Params,
) (*HilbertTransformerInstantaneousTrendLine, error) {
	return newHilbertTransformerInstantaneousTrendLine(
		p.EstimatorType, &p.EstimatorParams,
		p.AlphaEmaPeriodAdditional,
		p.TrendLineSmoothingLength,
		p.CyclePartMultiplier,
		p.BarComponent, p.QuoteComponent, p.TradeComponent)
}

//nolint:funlen,cyclop
func newHilbertTransformerInstantaneousTrendLine(
	estimatorType hilberttransformer.CycleEstimatorType,
	estimatorParams *hilberttransformer.CycleEstimatorParams,
	alphaEmaPeriodAdditional float64,
	trendLineSmoothingLength int,
	cyclePartMultiplier float64,
	bc entities.BarComponent, qc entities.QuoteComponent, tc entities.TradeComponent,
) (*HilbertTransformerInstantaneousTrendLine, error) {
	const (
		invalid    = "invalid hilbert transformer instantaneous trend line parameters"
		fmta       = "%s: α for additional smoothing should be in range (0, 1]"
		fmttlsl    = "%s: trend line smoothing length should be 2, 3, or 4"
		fmtcpm     = "%s: cycle part multiplier should be in range (0, 10]"
		fmtw       = "%s: %w"
		fmtn       = "htitl(%.3f, %d, %.3f%s%s)"
		fmtnDCP    = "dcp(%.3f%s%s)"
		four       = 4
		alpha      = 0.2
		descrValue = "Hilbert transformer instantaneous trend line "
		descrDCP   = "Dominant cycle period "
		tlslTwo    = 2
		tlslThree  = 3
		tlslFour   = 4
		cpmMax     = 10.0
	)

	if alphaEmaPeriodAdditional <= 0. || alphaEmaPeriodAdditional > 1. {
		return nil, fmt.Errorf(fmta, invalid)
	}

	if trendLineSmoothingLength < tlslTwo || trendLineSmoothingLength > tlslFour {
		return nil, fmt.Errorf(fmttlsl, invalid)
	}

	if cyclePartMultiplier <= 0. || cyclePartMultiplier > cpmMax {
		return nil, fmt.Errorf(fmtcpm, invalid)
	}

	// SineWave-style default: BarMedianPrice always shown in mnemonic.
	if bc == 0 {
		bc = entities.BarMedianPrice
	}

	if qc == 0 {
		qc = entities.DefaultQuoteComponent
	}

	if tc == 0 {
		tc = entities.DefaultTradeComponent
	}

	estimator, err := hilberttransformer.NewCycleEstimator(estimatorType, estimatorParams)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	estimatorMoniker := ""
	if estimatorType != hilberttransformer.HomodyneDiscriminator ||
		estimatorParams.SmoothingLength != four ||
		estimatorParams.AlphaEmaQuadratureInPhase != alpha ||
		estimatorParams.AlphaEmaPeriod != alpha {
		estimatorMoniker = hilberttransformer.EstimatorMoniker(estimatorType, estimator)
		if len(estimatorMoniker) > 0 {
			estimatorMoniker = ", " + estimatorMoniker
		}
	}

	componentMnemonic := core.ComponentTripleMnemonic(bc, qc, tc)

	mnemonic := fmt.Sprintf(fmtn, alphaEmaPeriodAdditional, trendLineSmoothingLength,
		cyclePartMultiplier, estimatorMoniker, componentMnemonic)
	mnemonicDCP := fmt.Sprintf(fmtnDCP, alphaEmaPeriodAdditional, estimatorMoniker, componentMnemonic)

	barFunc, err := entities.BarComponentFunc(bc)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	quoteFunc, err := entities.QuoteComponentFunc(qc)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	tradeFunc, err := entities.TradeComponentFunc(tc)
	if err != nil {
		return nil, fmt.Errorf(fmtw, invalid, err)
	}

	var c0, c1, c2, c3 float64

	switch trendLineSmoothingLength {
	case tlslTwo:
		c0, c1 = 2.0/3.0, 1.0/3.0
	case tlslThree:
		c0, c1, c2 = 3.0/6.0, 2.0/6.0, 1.0/6.0
	default: // tlslFour
		c0, c1, c2, c3 = 4.0/10.0, 3.0/10.0, 2.0/10.0, 1.0/10.0
	}

	maxPeriod := estimator.MaxPeriod()

	return &HilbertTransformerInstantaneousTrendLine{
		mnemonic:                       mnemonic,
		description:                    descrValue + mnemonic,
		mnemonicDCP:                    mnemonicDCP,
		descriptionDCP:                 descrDCP + mnemonicDCP,
		htce:                           estimator,
		alphaEmaPeriodAdditional:       alphaEmaPeriodAdditional,
		oneMinAlphaEmaPeriodAdditional: 1. - alphaEmaPeriodAdditional,
		cyclePartMultiplier:            cyclePartMultiplier,
		trendLineSmoothingLength:       trendLineSmoothingLength,
		coeff0:                         c0,
		coeff1:                         c1,
		coeff2:                         c2,
		coeff3:                         c3,
		input:                          make([]float64, maxPeriod),
		inputLength:                    maxPeriod,
		inputLengthMin1:                maxPeriod - 1,
		barFunc:                        barFunc,
		quoteFunc:                      quoteFunc,
		tradeFunc:                      tradeFunc,
	}, nil
}

// IsPrimed indicates whether an indicator is primed.
func (s *HilbertTransformerInstantaneousTrendLine) IsPrimed() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()

	return s.primed
}

// Metadata describes an output data of the indicator.
func (s *HilbertTransformerInstantaneousTrendLine) Metadata() core.Metadata {
	return core.BuildMetadata(
		core.HilbertTransformerInstantaneousTrendLine,
		s.mnemonic,
		s.description,
		[]core.OutputText{
			{Mnemonic: s.mnemonic, Description: s.description},
			{Mnemonic: s.mnemonicDCP, Description: s.descriptionDCP},
		},
	)
}

// Update updates the value of the indicator given the next sample, returning the
// (value, period) pair. Returns NaN values if the indicator is not yet primed.
func (s *HilbertTransformerInstantaneousTrendLine) Update(sample float64) (float64, float64) {
	if math.IsNaN(sample) {
		return sample, sample
	}

	s.mu.Lock()
	defer s.mu.Unlock()

	s.htce.Update(sample)
	s.pushInput(sample)

	if s.primed {
		s.smoothedPeriod = s.alphaEmaPeriodAdditional*s.htce.Period() +
			s.oneMinAlphaEmaPeriodAdditional*s.smoothedPeriod
		average := s.calculateAverage()
		s.value = s.coeff0*average + s.coeff1*s.average1 + s.coeff2*s.average2 + s.coeff3*s.average3
		s.average3 = s.average2
		s.average2 = s.average1
		s.average1 = average

		return s.value, s.smoothedPeriod
	}

	if s.htce.Primed() {
		s.primed = true
		s.smoothedPeriod = s.htce.Period()
		average := s.calculateAverage()
		s.value = average
		s.average1 = average
		s.average2 = average
		s.average3 = average

		return s.value, s.smoothedPeriod
	}

	nan := math.NaN()

	return nan, nan
}

// UpdateScalar updates the indicator given the next scalar sample.
func (s *HilbertTransformerInstantaneousTrendLine) UpdateScalar(sample *entities.Scalar) core.Output {
	return s.updateEntity(sample.Time, sample.Value)
}

// UpdateBar updates the indicator given the next bar sample.
func (s *HilbertTransformerInstantaneousTrendLine) UpdateBar(sample *entities.Bar) core.Output {
	return s.updateEntity(sample.Time, s.barFunc(sample))
}

// UpdateQuote updates the indicator given the next quote sample.
func (s *HilbertTransformerInstantaneousTrendLine) UpdateQuote(sample *entities.Quote) core.Output {
	return s.updateEntity(sample.Time, s.quoteFunc(sample))
}

// UpdateTrade updates the indicator given the next trade sample.
func (s *HilbertTransformerInstantaneousTrendLine) UpdateTrade(sample *entities.Trade) core.Output {
	return s.updateEntity(sample.Time, s.tradeFunc(sample))
}

func (s *HilbertTransformerInstantaneousTrendLine) updateEntity(t time.Time, sample float64) core.Output {
	const length = 2

	value, period := s.Update(sample)

	output := make([]any, length)
	output[0] = entities.Scalar{Time: t, Value: value}
	output[1] = entities.Scalar{Time: t, Value: period}

	return output
}

func (s *HilbertTransformerInstantaneousTrendLine) pushInput(value float64) {
	copy(s.input[1:], s.input[:s.inputLengthMin1])
	s.input[0] = value
}

func (s *HilbertTransformerInstantaneousTrendLine) calculateAverage() float64 {
	const half = 0.5

	// Compute the trend line as a simple average over the measured dominant cycle period.
	length := int(math.Floor(s.smoothedPeriod*s.cyclePartMultiplier + half))
	if length > s.inputLength {
		length = s.inputLength
	} else if length < 1 {
		length = 1
	}

	var sum float64
	for i := 0; i < length; i++ {
		sum += s.input[i]
	}

	return sum / float64(length)
}
